src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -316,15 +324,26 @@ public void Dispose()
         {
             lock (_sync)
             {
-                AbortAwaiting();
-
-                if (!_consuming)
+                if (!_disposed)
                 {
-                    ReturnBlocks(_head, null);
-                    _head = null;
-                    _tail = null;
+                    AbortAwaiting();
+
+                    if (!_consuming)
+                    {
+                        ReturnBlocks(_head, null);
+                        _head = null;
+                        _tail = null;
+                    }
+                    _disposed = true;
                 }
-                _disposed = true;
+            }
+        }
+
+        private void CheckDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(GetType().Name);
```

Comment:
`#if DEBUG`? While this shouldn't throw (but might in the case of ConnectionFilters), we probably shouldn't crash the entire server. If we should, we might as well put a `Trace.Assert` in the `MemoryPoolBlock` finalizer.

At least in the case of `ConnectionFilter`s, `IncommingData` will likely be called inside a `LoggingThreadPool` try/catch.